internal/stacks/stackruntime/internal/stackeval/stubs/offline.go

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package stubs
+
+import (
+	"github.com/hashicorp/terraform/internal/providers"
+	"github.com/hashicorp/terraform/internal/tfdiags"
+)
+
+// offlineProvider is a stub provider that is used in place of a provider that
+// is not configured  and should never be configured by the current Terraform
+// configuration.
+//
+// The only functionality that should be called on an offlineProvider are
+// provider function calls and move resource state.
+//
+// For everything else, Stacks should have provided a pre-configured provider
+// that should be used instead.
+type offlineProvider struct {
+	unconfiguredClient providers.Interface
+}
+
+func OfflineProvider(unconfiguredClient providers.Interface) providers.Interface {
+	return &offlineProvider{
+		unconfiguredClient: unconfiguredClient,
+	}
+}
+
+func (o *offlineProvider) GetProviderSchema() providers.GetProviderSchemaResponse {
+	// We do actually use the schema to work out which functions are available
+	// and whether cross-resource moves are even supported.
+	return o.unconfiguredClient.GetProviderSchema()
+}
+
+func (o *offlineProvider) ValidateProviderConfig(request providers.ValidateProviderConfigRequest) providers.ValidateProviderConfigResponse {
```

Comment:
While we probably won't call these right now, `Validate*` methods are offline, so maybe we shouldn't block them here?

Reply:
Potentially `offline` was a bad name for this provider as I was trying to be explicit about exactly which functions we expect to be called by this provider (which is just move and call function). If one of the validate functions is called then this provider is being used somewhere it shouldn't be, so just because it can be used in this way doesn't mean we should allow it.

We could always come back and re-enable the specific functions we need, when we discover that they are actually needed. As such, I'd lean towards leaving this as unavailable.

Reply:
gotcha, that makes sense. Looking at your doc comments for the type, it's spelled out right there 👍